umi_tools/umi_methods.py

```diff
@@ -1285,14 +1369,19 @@ def check_output(self):
         return do_output, out_keys
 
     def __call__(self, inreads):
+        if self.options.paired:
+            generator = get_readpairs()
+        else:
+            generator = get_singletons()
 
-        for read in inreads:
+        for read, read2 in generator(inreads):
```

Comment:
Since we are waiting for pairs to be output, can we any longer assume that reads will be output in the same order that they are in in the file. Specifically, if a mate is not found, it will not be output until the end? But in order to collect all reads that map to a location, we assume that they are sorted by pos.

Reply:
The pair iterator keeps the order of read1 intact -- it never moves past an incomplete pair in the output queue, but instead waits for that pair to be completed. get_bundles can thus till thrust to see reads in ascending (mapping) position of read1. The read2 positions will be out of order, but that shouldn't matter.

Reply:
This is now fixed, see commits below.